rasterio/session.py

```diff
@@ -0,0 +1,185 @@
+"""Abstraction for sessions in various clouds."""
+
+
+from rasterio.path import parse_path, UnparsedPath, ParsedPath
+
+
+class Session(object):
+    """Base for classes that configure access to secured resources.
+
+    Attributes
+    ----------
+    credentials : dict
+        Keys and values for session credentials.
+
+    Notes
+    -----
+    This class is not intended to be instantiated.
+
+    """
+
+    def get_credential_options(self):
+        """Get credentials as GDAL configuration options
+
+        Returns
+        -------
+        dict
+
+        """
+        return NotImplementedError
+
+    @staticmethod
+    def from_foreign_session(session, cls=None):
+        """Create a session object matching the foreign `session`.
+
+        Parameters
+        ----------
+        session : obj
+            A foreign session object.
+        cls : Session class, optional
+            The class to return.
+
+        Returns
+        -------
+        Session
+
+        """
+        if not cls:
+            return DummySession()
+        else:
+            return cls(session)
+
+    @staticmethod
+    def from_path(path, *args, **kwargs):
+        """Create a session object suited to the data at `path`.
+
+        Parameters
+        ----------
+        path : str
+            A dataset path or identifier.
+        args : sequence
+            Positional arguments for the foreign session constructor.
+        kwargs : dict
+            Keyword arguments for the foreign session constructor.
+
+        Returns
+        -------
+        Session
+
+        """
+        if not path:
+            return DummySession()
+
+        path = parse_path(path)
+
+        if isinstance(path, UnparsedPath) or path.is_local:
+            return DummySession()
+
+        elif path.scheme == "s3" or "amazonaws.com" in path.path:
+            return AWSSession(*args, **kwargs)
+
+        # This factory can be extended to other cloud providers here.
+        # elif path.scheme == "cumulonimbus":  # for example.
+        #     return CumulonimbusSession(*args, **kwargs)
+
+        else:
+            return DummySession()
+
+
+class DummySession(Session):
+    """A dummy session.
+
+    Attributes
+    ----------
+    credentials : dict
+        The session credentials.
+
+    """
+
+    def __init__(self, *args, **kwargs):
+        self._session = None
+        self.credentials = {}
+
+    def get_credential_options(self):
+        """Get credentials as GDAL configuration options
+
+        Returns
+        -------
+        dict
+
+        """
+        return {}
+
+
+class AWSSession(Session):
+    """Configures access to secured resources stored in AWS S3.
+    """
+
+    def __init__(
+            self, session=None, aws_unsigned=False, aws_access_key_id=None,
+            aws_secret_access_key=None, aws_session_token=None,
+            region_name=None, profile_name=None):
```

Comment:
@sgillies Could we add `kwargs` here to let the user pass other options like `aws_request_payer="requester"` ?

Reply:
@vincentsarago yes, I'll do it, but in a different PR.